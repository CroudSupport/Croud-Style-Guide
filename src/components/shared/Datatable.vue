<script>
    import debounce from 'lodash/debounce'
    import defaultsDeep from 'lodash/defaultsDeep'

    import Vuetable from 'vuetable-2/src/components/Vuetable'
    import VuetablePagination from 'vuetable-2/src/components/VuetablePaginationDropdown'
    import VuetablePaginationInfo from 'vuetable-2/src/components/VuetablePaginationInfo'

    /**
     * A flexible datatable that handles API calls and pagination
     * @example ./croud-datatable.md
     */
    export default {
        name: 'croud-datatable',

        components: {
            Vuetable,
            VuetablePagination,
            VuetablePaginationInfo,
        },

          /* eslint-disable object-shorthand */
        render: function (createElement) {
            return createElement('div', { class: 'ui basic segment' }, [
                createElement('vuetable', {
                    props: {
                        ...this.$props,
                        ...this.config,
                    },
                    scopedSlots: this.$scopedSlots,
                    ref: 'vuetable',
                    on: {
                        'vuetable:pagination-data': this.onPaginationData,
                        'vuetable:cell-clicked': this.onCellClicked,
                        'vuetable:loading': () => { this.loading = true },
                        'vuetable:loaded': () => { this.loading = false },
                    },
                }),
                createElement('div', { class: ['ui', 'segment', 'grid', 'basic'] }, [
                    createElement('vuetable-pagination-info', {
                        ref: 'paginationInfo',
                    }),
                    createElement('vuetable-pagination', {
                        ref: 'pagination',
                        on: {
                            'vuetable-pagination:change-page': this.onChangePage,
                        },
                    }),
                ]),
                createElement('div', {
                    class: {
                        ui: true,
                        inverted: true,
                        dimmer: true,
                        active: this.loading,
                    },
                }, [
                    createElement('div', { class: 'ui text large loader' }),
                ]),
            ])
        },

        props: {
            /**
             * vueTable-2 config that gets passed through to vue-table
             * @see https://github.com/ratiw/vuetable-2
             */
            vuetableConfig: {
                type: Object,
                required: true,
            },

            /**
             * Name of the globally registered component that will be used as the detail row
             * @see https://github.com/ratiw/vuetable-2-tutorial/wiki/lesson-12
             */
            detailRowComponent: {
                type: String,
            },

            /**
             * Transform function for preping the pagination data
             *
             * The default for this prop works with core
             */
            transform: {
                type: Function,
                default(data) {
                    data.meta.pagination = {
                        last_page: data.meta.pagination.total_pages,
                        next_page_url: data.meta.pagination.links.next,
                        prev_page_url: data.meta.pagination.links.previous,
                        from: (data.meta.pagination.per_page * data.meta.pagination.current_page) - (data.meta.pagination.per_page - 1),
                        to: data.meta.pagination.per_page * data.meta.pagination.current_page,
                        ...data.meta.pagination,
                    }
                    return data
                },
            },
        },

        methods: {
            onPaginationData(paginationData) {
                this.$emit('vuetable:pagination-data', paginationData)
                this.$refs.pagination.setPaginationData(paginationData)
                this.$refs.paginationInfo.setPaginationData(paginationData)
            },

            onChangePage(page) {
                this.$refs.vuetable.changePage(page)
            },

            onCellClicked(data) {
                this.$refs.vuetable.toggleDetailRow(data.id)
            },
        },

        data() {
            return {
                loading: false,
            }
        },

        computed: {
            config() {
                return defaultsDeep(this.vuetableConfig, {
                    'pagination-path': 'meta.pagination',
                    css: {
                        tableClass: 'ui table',
                        dropdownClass: 'ui dropdown',
                    },
                })
            },
        },

        watch: {
            vuetableConfig: {
                deep: true,
                handler: debounce(function () {
                    this.$refs.vuetable.refresh()
                }, 500),
            },
        },
    }
</script>
