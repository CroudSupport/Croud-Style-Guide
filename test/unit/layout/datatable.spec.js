import Vue from 'vue'
import axios from 'axios'
import Datatable from '../../../src/components/shared/Datatable'

const mock = jest.fn(() => Promise.resolve({
    data: {
        data: [
            { id: 3 },
            { id: 4 },
        ],
        meta: {
            pagination: {
                count: 2,
                current_page: 1,
                links: {},
                per_page: 15,
                total: 2,
                total_pages: 1,
            },
        },
    },
}))

axios.get = mock
const Constructor = Vue.extend(Datatable)

const vm = new Constructor({
    propsData: {
        vuetableConfig: {
            fields: ['id'],
            'api-url': 'test-url',
        },
    },
}).$mount()

const local = new Constructor({
    propsData: {
        vuetableConfig: {
            fields: ['id'],
            'api-mode': false,
            data: [{ id: 1 }, { id: 2 }],
        },
    },
}).$mount()

describe('Datatable', () => {
    describe('local data', () => {
        it('should not show pagination', () => {
            expect(Object.keys(local.$refs)).not.toContain('paginationInfo')
            expect(Object.keys(local.$refs)).not.toContain('pagination')
        })

        it('should match the snapshot', () => {
            expect(local.$el).toMatchSnapshot()
        })

        describe('no data', () => {
            it('should match the snapshot', (done) => {
                local.vuetableConfig.data = []
                setTimeout(() => {
                    expect(local.$el).toMatchSnapshot()
                    done()
                }, 1000)
            })
        })
    })

    describe('remote data', () => {
        it('should request remote data', () => {
            expect(mock.mock.calls).not.toBe(0)
        })

        it('should show pagination', () => {
            expect(Object.keys(vm.$refs)).toContain('paginationInfo')
            expect(Object.keys(vm.$refs)).toContain('pagination')
        })

        it('should match the snapshot', (done) => {
            vm.$nextTick(() => {
                expect(vm.$el).toMatchSnapshot()
                done()
            })
        })
    })
})
