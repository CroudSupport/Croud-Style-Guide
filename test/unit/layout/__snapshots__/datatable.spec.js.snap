// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datatable local data no data should match the snapshot 1`] = `
<div
  class="ui basic segment"
>
  <table
    class="vuetable ui table"
  >
    <thead>
      <tr>
        <th
          id="_id"
          class="vuetable-th-id"
        >
          Id
        </th>
      </tr>
    </thead>
    <tbody
      class="vuetable-body"
    >
      <tr>
        <td
          colspan="1"
          class="vuetable-empty-result"
        >
          <div
            class="ui center aligned basic segment"
          >
            <i
              class="big circular yellow list icon"
            />
            <br />
            No Results
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="ui inverted dimmer"
  >
    <div
      class="ui text large loader"
    />
  </div>
</div>
`;

exports[`Datatable local data should match the snapshot 1`] = `
<div
  class="ui basic segment"
>
  <table
    class="vuetable ui table"
  >
    <thead>
      <tr>
        <th
          id="_id"
          class="vuetable-th-id"
        >
          Id
        </th>
      </tr>
    </thead>
    <tbody
      class="vuetable-body"
    >
      <tr
        item-index="0"
        render="true"
        class=""
      >
        <td
          class=""
        >
          1
        </td>
      </tr>
      <!--  -->
      <tr
        item-index="1"
        render="true"
        class=""
      >
        <td
          class=""
        >
          2
        </td>
      </tr>
      <!--  -->
      <!--  -->
      <!--  -->
    </tbody>
  </table>
  <div
    class="ui inverted dimmer"
  >
    <div
      class="ui text large loader"
    />
  </div>
</div>
`;

exports[`Datatable remote data should match the snapshot 1`] = `
<div
  class="ui basic segment"
>
  <table
    class="vuetable ui table"
  >
    <thead>
      <tr>
        <th
          id="_id"
          class="vuetable-th-id"
        >
          Id
        </th>
      </tr>
    </thead>
    <tbody
      class="vuetable-body"
    >
      <tr
        item-index="0"
        render="true"
        class=""
      >
        <td
          class=""
        >
          3
        </td>
      </tr>
      <!--  -->
      <tr
        item-index="1"
        render="true"
        class=""
      >
        <td
          class=""
        >
          4
        </td>
      </tr>
      <!--  -->
      <!--  -->
      <!--  -->
    </tbody>
  </table>
  <div
    class="ui segment grid basic"
  >
    <div
      class="vuetable-pagination-info left floated left aligned six wide column"
    >
      Displaying 1 to 2 of 2 items
    </div>
    <!--  -->
  </div>
  <div
    class="ui inverted dimmer"
  >
    <div
      class="ui text large loader"
    />
  </div>
</div>
`;
